### PR TITLE
Flags

### PR DESCRIPTION
**Pull request by ramymedhat**
Adds a CLI to rbecc, avoiding the use of environment variables as flags.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -32,6 +33,18 @@ http_archive(
     ],
 )
 
+git_repository(
+    name   = "com_github_gflags_gflags",
+    #tag    = "v2.2.0",
```
Comment (gkousik) [inline WORKSPACE]: Remove?

File: src/main/cc/proxy_client/proxy_client.cc [hunk 2]
```diff
@@ -658,16 +647,26 @@ int ExecuteCommand(int argc, char** argv, const char** env) {
...
+      if (sep_idx == argc) {
+        cout << "You set the command to 'run'.\n";
+         cout << "Usage: " << gflags::ProgramInvocationShortName() <<
+              " [--arg=val] --command=run -- [full command]\n";
+         return 1;
+      }
+      return ExecuteCommand(argc-sep_idx-1, &argv[sep_idx+1], env);
+  } else if (FLAGS_command == "include_stats") {
+    return IncludeProcessorStats();
+  } else if (FLAGS_command == "list_includes") {
```
Comment (gkousik) [inline src/main/cc/proxy_client/proxy_client.cc]: @ola-rozenfeld I think we added list_includes as a debug end-point to make sure it was working in the beginning? Now that we know its fairly stable, can we remove it?

**Review by gkousik — COMMENTED** [review]
1. Did you test local fallback after these flag changes? Does execvp() still work as expected?
2. I think merge of this change should be co-ordinated with changes to ninja and our scripts? If so can we merge it on Monday and all of us update our tools at the same time?

Otherwise, the changes itself LGTM.

File: src/main/cc/proxy_client/proxy_client.cc (hunk 2 above)
Comment (ola-rozenfeld) [reply src/main/cc/proxy_client/proxy_client.cc]: I don't know, I actually like that we have an include_processor client as well. Whether that really needs to be the same binary as the proxy client is a good question.

File: src/main/cc/proxy_client/proxy_client.cc [hunk 3]
```diff
@@ -33,6 +34,34 @@
...
+DEFINE_bool(allow_out_under_in, false, "Indicate whether to allow outputs under"
+            " input directories");
+DEFINE_bool(allow_out_dirs, false, "Indicate whether to allow output directories"
+            " as inputs");
+DEFINE_string(out_dir, "out", "Name of the output directory");
+DEFINE_string(invocation_id, "", "Invocation ID of the build");
+DEFINE_int32(retry_count, 1, "Number of remote retries to attempt");
+DEFINE_int32(proxy_instances, 1, "Number of remote client proxy instances");
+DEFINE_string(proxy_address, "localhost:8080", "Address of the remote client proxy");
+DEFINE_int32(server_instances, 1, "Number of include processor server instances");
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc]: If you want a shorter flag name, go with ip_server*. But I'll rather have include_processor_server*. The convention is to use readable names for flags.

File: src/main/cc/proxy_client/proxy_client.cc [hunk 4]
```diff
@@ -33,6 +34,34 @@
...
+DEFINE_string(server_address, "localhost:8070", "Address of the include processor server");
+DEFINE_string(command, "", "Type of command: run, list_includes, include_stats");
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc:54]: I wonder if the C++ gflags has anything for this. The Py version, for example, has the appcommand module, to define command-specific flags.

File: src/main/cc/proxy_client/proxy_client.cc [hunk 5]
```diff
@@ -434,30 +448,19 @@ int CreateRunRequest(int argc, char** argv, const char** env,
...
-  if (accept_cached != nullptr) {
-    req->add_command("--accept_cached");
-    req->add_command(accept_cached);
-  }
-  char* save_exec_data = getenv("SAVE_EXECUTION_DATA");
-  if (save_exec_data != nullptr) {
-    req->add_command("--save_execution_data");
-    req->add_command(save_exec_data);
-  }
+  req->add_command("--invocation_id");
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc]: I think you still need to check whether it's empty.

File: src/main/cc/proxy_client/proxy_client.cc [hunk 6]
```diff
@@ -517,8 +520,15 @@ int CreateRunRequest(int argc, char** argv, const char** env,
     string varval(*env++);
     unsigned int eq_index = varval.find("=");
     string var = varval.substr(0, eq_index);
-    if (var == "PWD")
+    if (var == "PWD") {
       absl::StrAppend(&env_vars, varval, ",");
+      continue;
+    }
+    for (const auto& whitelisted_var : absl::StrSplit(FLAGS_env_whitelist, ',', absl::SkipEmpty())) {
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc]: Don't do that every time, read this into a set once, and then do set lookups for every variable. Also, just add PWD to the set to begin with.

File: src/main/cc/proxy_client/proxy_client.cc (hunk 2 above)
Comment (gkousik) [reply src/main/cc/proxy_client/proxy_client.cc]: Right ok, maybe it makes sense to keep it for now and remove when we are cleaning up rbecc.

File: src/main/cc/proxy_client/proxy_client.cc [hunk 7]
```diff
@@ -658,16 +647,26 @@ int ExecuteCommand(int argc, char** argv, const char** env) {
...
+  gflags::ParseCommandLineFlags(&sep_idx, &argv, false);
+  if (FLAGS_command == "run") {
+      if (sep_idx == argc) {
+        cout << "You set the command to 'run'.\n";
+         cout << "Usage: " << gflags::ProgramInvocationShortName() <<
+              " [--arg=val] --command=run -- [full command]\n";
+         return 1;
+      }
+      return ExecuteCommand(argc-sep_idx-1, &argv[sep_idx+1], env);
+  } else if (FLAGS_command == "include_stats") {
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc]: No need for else, every if has a return.

File: src/main/cc/proxy_client/proxy_client.cc (hunk 6 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc]: Done.

File: src/main/cc/proxy_client/proxy_client.cc (hunk 5 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc]: Done

File: src/main/cc/proxy_client/proxy_client.cc (hunk 4 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc:54]: Doesn't look like it. https://gflags.github.io/gflags/

File: src/main/cc/proxy_client/proxy_client.cc (hunk 3 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc]: Done.

File: WORKSPACE (hunk 1 above)
Comment (ramymedhat) [reply WORKSPACE]: Done

**Comment by ramymedhat** [comment]
> 1. Did you test local fallback after these flag changes? Does execvp() still work as expected?
> 2. I think merge of this change should be co-ordinated with changes to ninja and our scripts? If so can we merge it on Monday and all of us update our tools at the same time?
> 
> Otherwise, the changes itself LGTM.

1. Yes, tested and works fine.
2. Sure.

File: src/main/cc/proxy_client/proxy_client.cc (hunk 7 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc]: Done

File: src/main/cc/proxy_client/proxy_client.cc [hunk 8]
```diff
@@ -512,13 +518,15 @@ int CreateRunRequest(int argc, char** argv, const char** env,
   req->add_command("\\.o-.*$");
   req->add_command("\\.git.*$");
   req->add_command("--environment_variables");
-  string env_vars;
+  string env_vars = "PWD=" + kPWDOverride + ",";
+  std::set<std::string> whitelist = absl::StrSplit(FLAGS_env_whitelist, ',', absl::SkipEmpty());
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc]: No need for std:: on both set and string.

File: src/main/cc/proxy_client/proxy_client.cc [hunk 9]
```diff
@@ -512,13 +518,15 @@ int CreateRunRequest(int argc, char** argv, const char** env,
   req->add_command("\\.o-.*$");
   req->add_command("\\.git.*$");
   req->add_command("--environment_variables");
-  string env_vars;
+  string env_vars = "PWD=" + kPWDOverride + ",";
```
Comment (ola-rozenfeld) [inline src/main/cc/proxy_client/proxy_client.cc:518]: Are you sure you want to do the override here? I think it's much cleaner to do this outside, as we used to, and just always call this with PWD set to /proc/self/cwd and in the whitelist. This /proc/self/cwd thing is very Android-specific. But it can be a separate CL to take it out, if you prefer.

Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc:518]: I thought about this. I actually think pwd is not android specific. We can't do a deterministic compilation without it, and relying on someone else setting it is unnecessary imo. I also think bazel does this by default for linux (https://github.com/bazelbuild/bazel/issues/1000#issuecomment-285344868)

File: src/main/cc/proxy_client/proxy_client.cc (hunk 8 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc]: Done

File: src/main/cc/proxy_client/proxy_client.cc (hunk 9 above)
Comment (ramymedhat) [reply src/main/cc/proxy_client/proxy_client.cc:518]: Also note that we always did it within rbecc: https://github.com/ola-rozenfeld/tools_remote/blob/tool/src/main/cc/proxy_client/proxy_client.cc#L589

**Review by ola-rozenfeld — APPROVED** [review]
